sdks/python/apache_beam/runners/portability/universal_local_runner.py

```diff
@@ -0,0 +1,398 @@
+
+      else:
+        self._handle = JobServicer(use_grpc=False)
+
+    return self._handle
+
+  def run(self, pipeline):
+    handle = self._get_handle()
+    prepare_response = handle.Prepare(
+        beam_job_api_pb2.PrepareJobRequest(
+            job_name='job',
+            pipeline=pipeline.to_runner_api()))
+    run_response = handle.Run(beam_job_api_pb2.RunJobRequest(
+        preparation_id=prepare_response.preparation_id))
+    return PipelineResult(handle, run_response.job_id)
+
+
+class PipelineResult(runner.PipelineResult):
+  def __init__(self, handle, job_id):
+    super(PipelineResult, self).__init__(beam_job_api_pb2.JobState.UNKNOWN)
+    self._handle = handle
+    self._job_id = job_id
+    self._messages = []
+
+  def cancel(self):
+    self._handle.Cancel()
+
+  @property
+  def state(self):
+    runner_api_state = self._handle.GetState(
+        beam_job_api_pb2.GetJobStateRequest(job_id=self._job_id)).state
+    self._state = self._runner_api_state_to_pipeline_state(runner_api_state)
+    return self._state
+
+  @staticmethod
+  def _runner_api_state_to_pipeline_state(runner_api_state):
+    return getattr(
+        runner.PipelineState,
+        beam_job_api_pb2.JobState.JobStateType.Name(runner_api_state))
+
+  @staticmethod
+  def _pipeline_state_to_runner_api_state(pipeline_state):
+    return beam_job_api_pb2.JobState.JobStateType.Value(pipeline_state)
+
+  def wait_until_finish(self):
+    def read_messages():
+      for message in self._handle.GetMessageStream(
+          beam_job_api_pb2.JobMessagesRequest(job_id=self._job_id)):
+        self._messages.append(message)
+    threading.Thread(target=read_messages).start()
+
+    for state_response in self._handle.GetStateStream(
+        beam_job_api_pb2.GetJobStateRequest(job_id=self._job_id)):
+      self._state = self._runner_api_state_to_pipeline_state(
+          state_response.state)
+      if state_response.state in TERMINAL_STATES:
+        break
+    if self._state != runner.PipelineState.DONE:
+      raise RuntimeError(
+          "Pipeline %s failed in state %s." % (self._job_id, self._state))
+
+
+class BeamJob(threading.Thread):
+  """This class handles running and managing a single pipeline.
+  """
+  def __init__(self, job_id, pipeline_options, pipeline_proto,
+               use_grpc=True, sdk_harness_factory=None):
+    super(BeamJob, self).__init__()
+    self._job_id = job_id
+    self._pipeline_options = pipeline_options
+    self._pipeline_proto = pipeline_proto
+    self._use_grpc = use_grpc
+    self._sdk_harness_factory = sdk_harness_factory
+    self._log_queue = queue.Queue()
+    self._state_change_callbacks = [
+        lambda new_state: self._log_queue.put(
+            beam_job_api_pb2.JobMessagesResponse(
+                state_response=
+                beam_job_api_pb2.GetJobStateResponse(state=new_state)))
+    ]
+    self._state = None
+    self.state = beam_job_api_pb2.JobState.STARTING
```

Comment:
<!--new_thread; commit:77e51af3ee2af39264c366a2602957e7e9bb20c7; resolved:0-->
Since this is a public attribute, let's document it in class docstring.

Reply:
Done.